src/util/AsioHelpers.h

```diff
@@ -15,14 +15,26 @@ namespace net = boost::asio;
 
 /// Helper function that ensures that co_await resumes on the same
 /// executor it started with.
-/// IMPORTANT: If the inner awaitable throws, no guarantees are given. Make
-/// sure to keep that in mind when handling errors inside coroutines!
+/// IMPORTANT: If the coroutine is cancelled, no guarantees are given. Make
+/// sure to keep that in mind when handling cancellation errors!
 template <typename T>
 inline net::awaitable<T> sameExecutor(net::awaitable<T> awaitable) {
   auto initialExecutor = co_await net::this_coro::executor;
-  T result = co_await std::move(awaitable);
-  co_await net::dispatch(initialExecutor, net::use_awaitable);
-  co_return result;
+  std::exception_ptr exceptionPtr;
+  try {
+    T result = co_await std::move(awaitable);
+    co_await net::dispatch(initialExecutor, net::use_awaitable);
+    co_return result;
+  } catch (...) {
+    exceptionPtr = std::current_exception();
+  }
+  auto cancellationState = co_await net::this_coro::cancellation_state;
+  if (cancellationState.cancelled() == net::cancellation_type::none) {
+    co_await net::dispatch(initialExecutor, net::use_awaitable);
+  }
+  if (exceptionPtr) {
+    std::rethrow_exception(exceptionPtr);
+  }
```

Comment:
```suggestion
  AD_CORRECTNESS_CHECK(exceptionPtr);
  std::rethrow_exception(exceptionPtr);
```

This should get rid of the Warnings in Clang 16